pkg/foxtrot/jwt.go

```diff
@@ -0,0 +1,73 @@
+package foxtrot
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"foxygo.at/s/errs"
+)
+
+var (
+	encodedJWTHeader = jsonBase64Encode(map[string]string{"alg": "HS256"})
+
+	errJWT          = errors.New("invalid JWT Token")
+	errJWTEncoding  = fmt.Errorf("%w: bad encoding", errJWT)
+	errJWTExpired   = fmt.Errorf("%w: expired", errJWT)
+	errJWTSignature = fmt.Errorf("%w: invalid signature", errJWT)
+)
+
+type jwtPayload struct {
+	Sub string `json:"sub"` // subject: user name
+	Exp int64  `json:"exp"` // expiry in unix epoche seconds
+}
+
+func validateJWT(jwt string, secret []byte) error {
+	parts := strings.Split(jwt, ".")
+	if len(parts) != 3 {
+		return errs.Errorf("%v: invalid format, expected two '.' got %d", errJWT, len(parts))
+	}
+	if parts[0] != encodedJWTHeader {
+		return errs.Errorf("%v: invalid JWT header", errJWT)
+	}
+	payload := parts[1]
+	b, err := base64.RawURLEncoding.DecodeString(payload)
+	if err != nil {
+		return errs.Errorf("%v: base64: %v", errJWTEncoding, err)
+	}
+	p := jwtPayload{}
+	if err := json.Unmarshal(b, &p); err != nil {
+		return errs.Errorf("%v: json: %v", errJWTEncoding, err)
+	}
+	expectedSignature := sign(encodedJWTHeader+"."+payload, secret)
```

Comment:
I would check the signature before decoding the payload. This prevents you operating on data provided by an attacker that is not valid. I would also use `parts[0]` here instead of `encodedJWTHeader` because that is what came in from the outside. I know you check it for equality, but in future, that check could change so you would then need to change this anyway.